### PR TITLE
Add separate CI job for slow tests

### LINKED ISSUE #9086: Slow CI
## 🚀 Feature

Implement "Slow CI" jobs

### Motivation

Some tests spawn multiple works and can take 5-30 seconds to run. It's important that we run these but they can really become a bottleneck.

To avoid that, create a new CI job for them.

### Pitch

```python
@RunIf(slow=True)
def test_thing():
    ...
```

Would be run on latest Python with LTS PyTorch on Windows, Linux, Mac (3 jobs).
CPU only.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Just 1 parametrization for the moment, we can always add more later if necessary.

If I forgot to mark a specific test as `slow=True`, feel free to commit or comment.

Fixes #9086

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @carmocca @akihironitta @borda

### CODE REVIEW

File: .github/workflows/ci_test-slow.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      uses: codecov/codecov-action@v1
+      if: always()
+      # see: https://github.com/actions/toolkit/issues/399
+      continue-on-error: true
+      with:
+        token: ${{ secrets.CODECOV_TOKEN }}
+        file: coverage.xml
+        flags: cpu,pytest,torch${{ matrix.pytorch-version }}
+        name: CPU-coverage
+        fail_ci_if_error: false
```
Comment (carmocca) [inline .github/workflows/ci_test-slow.yml:82]: @Borda I'd appreciate it if you can check the coverage and cache steps as I'm not very familiar with their specifics.

File: tests/utilities/test_auto_restart.py [hunk 2]
```diff
@@ -897,10 +895,7 @@ def _run_training(trainer_kwargs, dataset_classes, fail_on_step: int = -1, ckpt_
     return model.seen_batches, model.parameters()
 
 
-# this test will fail `fault_tolerant` don't support multiple datasets.
-# this tests works as the dataset is fully deterministic and therefore
-# there is not overall between the seeds.
```
Comment (carmocca) [inline tests/utilities/test_auto_restart.py:902]: Removed this comment because I did not understand what it meant to say. Can be added back if somebody knows

File: .github/workflows/ci_test-slow.yml (hunk 1 above)
Comment (Borda) [reply .github/workflows/ci_test-slow.yml:82]: looks good, I would just upload coverage if succeeded...

File: .github/workflows/ci_test-slow.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  slow:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        # same config as '.azure-pipelines/gpu-tests.yml'
+        python-version: ["3.7"]
+        pytorch-version: ["1.8"]
+
+    timeout-minutes: 20
```
Comment (Borda) [inline .github/workflows/ci_test-slow.yml:20]: only 20min? sounds fast :]

File: .github/workflows/ci_test-slow.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    - name: Weekly reset caching
+      run: echo "::set-output name=period::$(python -c 'import time ; days = time.time() / 60 / 60 / 24 ; print(int(days / 7))' 2>&1)"
+      id: times
+
+    - name: Get pip cache
+      id: pip-cache
+      run: |
+        python -c "from pip._internal.locations import USER_CACHE_DIR; print('::set-output name=dir::' + USER_CACHE_DIR)"
+
+    - name: Cache pip
```
Comment (Borda) [inline .github/workflows/ci_test-slow.yml:36]: let's try the new caching? but that won't have the weekly reset...

File: .github/workflows/ci_test-slow.yml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        pip install --requirement requirements.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html --upgrade
+        pip install --requirement requirements/test.txt
+        pip list
+      shell: bash
+
+    - name: Tests
+      run: |
+        coverage run --source pytorch_lightning -m pytest tests -v --junitxml=junit/test-results-${{ runner.os }}-${{ matrix.python-version }}.xml
+      env:
+        PL_RUN_SLOW_TESTS: 1
```
Comment (Borda) [inline .github/workflows/ci_test-slow.yml:57]: thinking what it means? is it slow only or also slow...
how a user would all locally?

File: .github/workflows/ci_test-slow.yml (hunk 3 above)
Comment (carmocca) [reply .github/workflows/ci_test-slow.yml:20]: Because we don't have many "slow tests at the moment". I think this could be tuned lower

File: .github/workflows/ci_test-slow.yml (hunk 4 above)
Comment (carmocca) [reply .github/workflows/ci_test-slow.yml:36]: Let's do that in a separate PR for all jobs.

File: .github/workflows/ci_test-slow.yml (hunk 5 above)
Comment (carmocca) [reply .github/workflows/ci_test-slow.yml:57]: Slow only, works the same way as standalone.

> how a user would all locally?

Can't, same as standalone

File: .github/workflows/ci_test-slow.yml (hunk 1 above)
Comment (carmocca) [reply .github/workflows/ci_test-slow.yml:82]: Do you mean to remove `continue_on_error: true`?

Comment (justusschock) [reply .github/workflows/ci_test-slow.yml:82]: probably change the `if: always()` to `if: success()`

File: .github/workflows/ci_test-slow.yml (hunk 3 above)
Comment (carmocca) [reply .github/workflows/ci_test-slow.yml:21]: Ubuntu took 16 min and mac 13. Seems like 20 is good for now

File: tests/helpers/runif.py [hunk 6]
```diff
@@ -195,6 +197,13 @@ def __new__(
             conditions.append(not _OMEGACONF_AVAILABLE)
             reasons.append("omegaconf")
 
+        if slow:
```
Comment (awaelchli) [inline tests/helpers/runif.py:200]: a job cannot be marked as slow and require gpu. should we handle this case? There may be other collisions which might be annoying.

**Review by awaelchli — APPROVED** [review]

File: tests/helpers/runif.py (hunk 6 above)
Comment (carmocca) [reply tests/helpers/runif.py:200]: We can do a simple `if` check for it here if both are passed together, but it won't work with:

```python
@RunIf(min_gpus=1)
@RunIf(slow=True)
def test_something():
    ...

# or, more common pattern

@RunIf(slow=True)
@pytest.mark.parametrize("an_arg", [False, pytest.param(True, marks=RunIf(min_gpus=1))])
def test_something(an_arg):
    ...
```

File: .github/workflows/ci_test-slow.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        pytorch-version: ["1.8"]
+
+    timeout-minutes: 20
+    steps:
+    - uses: actions/checkout@v2
+    - uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Weekly reset caching
```
Comment (tchaton) [inline .github/workflows/ci_test-slow.yml:28]: It is required ? Why weekly ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: .github/workflows/ci_test-slow.yml (hunk 7 above)
Comment (carmocca) [reply .github/workflows/ci_test-slow.yml:28]: Copying it from other test jobs. cc @Borda 

We plan to re-evaluate caching, but in another PR

File: .github/workflows/ci_test-slow.yml (hunk 1 above)
Comment (Borda) [reply .github/workflows/ci_test-slow.yml:83]: I would go with `if: success()`

**Review by Borda — APPROVED** [review]

**Comment by awaelchli** [comment]
Not sure if it was discussed, but was it intentional to skip slow tests by default? This means locally all tests marked as slow don't run. Since locally we usually run only a subset of tests anyway, "slowness" doesn't play such a big role. May I suggest to invert the current behavior?

**Comment by carmocca** [comment]
Skipping them by default is the point of adding Slow CI, otherwise, they get to run on all jobs and we get no speedup.

**Comment by awaelchli** [comment]
You would make the default to run the slow tests and set

PL_RUN_SLOW_TESTS=0 in the regular CI and 
PL_RUN_SLOW_TESTS=1 in the dedicated one.

Currently only the slow ci configuration sets the env variable.

**Comment by carmocca** [comment]
We would need 3 values:

1. Run only slow tests (current `PL_RUN_SLOW_TESTS=1`)
1. Run only non-slow tests (current `PL_RUN_SLOW_TESTS=0`)
1. Run all tests. How do you propose we set it?

File: .github/workflows/ci_test-slow.yml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  pull_request:
+    branches: [master, "release/*"]
+
+jobs:
+  slow:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macOS-latest]
```
Comment (Borda) [inline .github/workflows/ci_test-slow.yml:16]: just noticed that here we skip run on Windows and all minimal requirements?

Comment (carmocca) [reply .github/workflows/ci_test-slow.yml:16]: Yes, I wanted to keep this light at first as we don't run many tests at the moment. These can be added later
